File: tosctl/src/validator-auth-native/src/native_evidence.rs
use crate::{cells, native, native_history::NativeFinalizedHistory};
use chain_block::{BuilderData, Cell, CellType, HashmapE, HashmapType, SliceData};
use std::{collections::BTreeMap, sync::Arc};
use tos_validator_auth::{
    codec::{decode, Error, Hash},
    transfer::{validate_object, ObjectReader, CHUNK_BYTES, INLINE_BYTES},
    types::{Anchor, Authorizations, ObjectRef},
};
pub const NATIVE_EVIDENCE_TAG: u32 = 0x76616531;
pub const NATIVE_AUTHORIZATIONS_LIMIT: usize = 4 * INLINE_BYTES + 4096;
type ChunkKey = [u8; 33];
fn need(ok: bool, code: &'static str) -> Result<(), Error> {
    if ok {
        Ok(())
    } else {
        Err(Error(code))
    }
}
fn key(id: &Hash, index: u8) -> ChunkKey {
    let mut result = [0; 33];
    result[..32].copy_from_slice(id);
    result[32] = index;
    result
}
fn bits(k: &ChunkKey) -> Result<SliceData, Error> {
    native(SliceData::load_bitstring(native(BuilderData::with_raw(k.to_vec(), 264))?))
}
fn declared(cell: Cell, limit: usize) -> Result<usize, Error> {
    need(
        cell.cell_type() == CellType::Ordinary
            && cell.level() == 0
            && cell.bit_length() == 336
            && cell.references_count() == 1,
        "auth-bytes-shape",
    )?;
    let mut s = native(SliceData::load_cell(cell))?;
    need(
        native(s.get_next_u32())? == 0x76616231 && native(s.get_next_u16())? == 1,
        "auth-bytes-version",
    )?;
    let n = native(s.get_next_u32())? as usize;
    need(n != 0 && n <= limit, "evidence-byte-bound")?;
    Ok(n)
}
/// A transaction-contained set of typed objects. Construction does not confer
/// owner, possession, identity or governance authority.
pub struct NativeEvidence {
    authorizations: Authorizations,
    header: Cell,
    chunks: Arc<BTreeMap<ChunkKey, Vec<u8>>>,
}
impl NativeEvidence {
    /// Charge deterministic byte bounds before decoding/hashing them. Native
    /// execution chooses gas prices; no IO or cache affects these charges.
    pub fn open(
        root: Cell,
        mut charge: impl FnMut(usize) -> Result<(), Error>,
    ) -> Result<Self, Error> {
        need(
            root.cell_type() == CellType::Ordinary
                && root.level() == 0
                && root.bit_length() == 49
                && root.references_count() >= 2,
            "evidence-shape",
        )?;
        let mut s = native(SliceData::load_cell(root))?;
        need(
            native(s.get_next_u32())? == NATIVE_EVIDENCE_TAG && native(s.get_next_u16())? == 1,
            "evidence-version",
        )?;
        let present = native(s.get_next_bit())?;
        need(s.remaining_references() == 2 + usize::from(present), "evidence-shape")?;
        let auth = native(s.checked_drain_reference())?;
        let header = native(s.checked_drain_reference())?;
        charge(declared(auth.clone(), NATIVE_AUTHORIZATIONS_LIMIT)?)?;
        let a: Authorizations = decode(&cells::unpack_bytes(auth, NATIVE_AUTHORIZATIONS_LIMIT)?)?;
        if a.owner.is_empty() {
            need(
                header.cell_type() == CellType::Ordinary
                    && header.bit_length() == 0
                    && header.references_count() == 0,
                "evidence-unused-header",
            )?;
        }
        let objects: Vec<_> = a
            .owner
            .iter()
            .map(|v| (&v.proof.proof, 5))
            .chain(a.administration.iter().map(|v| (&v.certificate, 4)))
            .chain(a.governance.iter().map(|v| (&v.certificate, 4)))
            .collect();
        let mut expected = BTreeMap::new();
        let mut manifests = BTreeMap::new();
        let mut remaining: usize = 67_108_864;
        for &(v, kind) in &objects {
            validate_object(v, kind)?;
            let size = if v.inline.is_empty() {
                v.reference[0].byte_length as usize
            } else {
                v.inline.len()
            };
            remaining = remaining.checked_sub(size).ok_or(Error("attachment-budget"))?;
            let Some(r) = v.reference.first() else { continue };
            if let Some(old) = manifests.insert(r.object_id, r) {
                need(old == r, "evidence-manifest-conflict")?;
            }
            for i in 0..r.chunk_hashes.len() {
                let length = (r.byte_length as usize)
                    .checked_sub(i.checked_mul(CHUNK_BYTES).ok_or(Error("chunk-length"))?)
                    .ok_or(Error("chunk-length"))?
                    .min(CHUNK_BYTES);
                expected.insert(
                    key(&r.object_id, u8::try_from(i).map_err(|_| Error("chunk-index"))?),
                    length,
                );
            }
        }
        // Only admitted manifest keys may be enumerated; unused entries fail
        // before their payload is decoded or recursively traversed.
        let supplied = HashmapE::with_hashmap(
            264,
            if present { Some(native(s.checked_drain_reference())?) } else { None },
        );
        let mut canonical = HashmapE::with_bit_len(264);
        let mut chunks_cells = BTreeMap::new();
        let mut failure = None;
        let complete = native(supplied.iterate_slices(|k, v| {
            let checked = (|| -> Result<(), Error> {
                need(
                    k.remaining_bits() == 264
                        && v.remaining_bits() == 0
                        && v.remaining_references() == 1,
                    "evidence-chunk-shape",
                )?;
                let k: ChunkKey =
                    k.get_bytestring(0).try_into().map_err(|_| Error("evidence-chunk-shape"))?;
                let n = expected.get(&k).ok_or(Error("evidence-extra-chunk"))?;
                let cell = native(v.reference(0))?;
                need(declared(cell.clone(), CHUNK_BYTES)? == *n, "evidence-chunk-length")?;
                chunks_cells.insert(k, cell.clone());
                need(native(canonical.setref(bits(&k)?, cell))?.is_none(), "evidence-dictionary")?;
                Ok(())
            })();
            match checked {
                Ok(()) => Ok(true),
                Err(e) => {
                    failure = Some(e);
                    Ok(false)
                }
            }
        }))?;
        if let Some(e) = failure {
            return Err(e);
        }
        need(complete, "evidence-dictionary")?;
        need(chunks_cells.len() == expected.len(), "evidence-missing-chunk")?;
        need(
            supplied.data().map(Cell::repr_hash) == canonical.data().map(Cell::repr_hash),
            "evidence-dictionary",
        )?;
        charge(67_108_864_usize.checked_sub(remaining).ok_or(Error("attachment-budget"))?)?;
        let mut chunks = BTreeMap::new();
        for (k, cell) in chunks_cells {
            chunks.insert(
                k,
                cells::unpack_bytes(cell, *expected.get(&k).ok_or(Error("evidence-extra-chunk"))?)?,
            );
        }
        let result = Self { authorizations: a.clone(), header, chunks: Arc::new(chunks) };
        let mut reader = result.reader();
        for (v, kind) in objects {
            reader.resolve(v, kind)?;
        }
        Ok(result)
    }
    pub fn authorizations(&self) -> &Authorizations {
        &self.authorizations
    }
    pub fn reader(&self) -> ObjectReader<impl FnMut(&ObjectRef, u8) -> Result<Vec<u8>, Error>> {
        let chunks = self.chunks.clone();
        ObjectReader::new(move |r: &ObjectRef, i: u8| {
            chunks.get(&key(&r.object_id, i)).cloned().ok_or(Error("evidence-missing-chunk"))
        })
    }
    pub fn authenticate_owner<F>(
        &self,
        history: &NativeFinalizedHistory<F>,
    ) -> Result<Anchor, Error>
    where
        F: Fn(&chain_block::BlockIdExt, usize) -> Result<Vec<u8>, Error>,
    {
        need(self.authorizations.owner.len() == 1, "evidence-owner-absent")?;
        let expected = &self.authorizations.owner[0].proof.anchor;
        let actual = history.authenticate_header(expected.seqno, self.header.clone())?;
        need(&actual == expected, "evidence-owner-anchor")?;
        Ok(actual)
    }
}

File: validator/auth/native-evidence.cpp
#include "vm/cells/CellBuilder.h"
#include "vm/cells/CellSlice.h"
#include "vm/dict.h"

#include "native-evidence.h"
namespace tos::auth {
namespace {
void need(bool ok, const char* code) {
  if (!ok)
    throw Error{code};
}
template <class T>
T take(Result<T> r) {
  if (!r.ok())
    throw r.error();
  return std::move(r.value());
}
using ChunkKey = std::array<std::uint8_t, 33>;
ChunkKey key(const Hash& id, std::uint8_t index) {
  ChunkKey k{};
  std::copy(id.begin(), id.end(), k.begin());
  k[32] = index;
  return k;
}
std::size_t declared(td::Ref<vm::Cell> root, std::size_t limit) {
  vm::CellSlice s{vm::NoVm{}, root};
  need(s.is_valid() && !s.is_special() && root->get_level() == 0 && s.size() == 336 && s.size_refs() == 1,
       "auth-bytes-shape");
  need(s.fetch_ulong(32) == 0x76616231 && s.fetch_ulong(16) == 1, "auth-bytes-version");
  auto n = s.fetch_ulong(32);
  need(n != 0 && n <= limit, "evidence-byte-bound");
  return n;
}
}  // namespace
ObjectReader NativeEvidence::reader() const {
  return ObjectReader([chunks = chunks_](const ObjectRef& ref, std::uint8_t index) -> Result<Bytes> {
    auto it = chunks->find(key(ref.object_id_, index));
    // Admission already required every declared chunk to be present, so a
    // resolve that misses one would mean a reference the manifest never
    // declared. No case in the evidence suite reaches this: relabelling it
    // leaves all of them passing. It stays because removing a refusal whose
    // condition is merely believed unreachable is how the belief stops being
    // checked, but it is not a guard anything currently proves.
    if (it == chunks->end())
      return Error{"evidence-missing-chunk"};
    return it->second;
  });
}
Result<NativeEvidence> NativeEvidence::open(td::Ref<vm::Cell> root, const EvidenceCharge& charge) {
  try {
    need(bool(charge), "evidence-charge");
    vm::CellSlice s{vm::NoVm{}, root};
    need(s.is_valid() && !s.is_special() && root->get_level() == 0 && !root->is_virtualized() && s.size() == 49 &&
             s.size_refs() >= 2,
         "evidence-shape");
    need(s.fetch_ulong(32) == native_evidence_tag && s.fetch_ulong(16) == 1, "evidence-version");
    bool present = s.fetch_ulong(1);
    need(s.size_refs() == 2 + unsigned(present), "evidence-shape");
    auto auth = s.fetch_ref();
    auto header = s.fetch_ref();
    auto bytes = declared(auth, native_authorizations_limit);
    need(take(charge(bytes)), "evidence-charge");
    auto a = take(decode<Authorizations>(take(unpack_bytes(auth, native_authorizations_limit))));
    if (a.owner_.empty()) {
      vm::CellSlice h{vm::NoVm{}, header};
      need(!h.is_special() && h.size() == 0 && h.size_refs() == 0, "evidence-unused-header");
    }
    std::vector<std::pair<const ObjectValue*, std::uint8_t>> objects;
    for (const auto& v : a.owner_)
      objects.emplace_back(&v.proof_.proof_, 5);
    for (const auto& v : a.administration_)
      objects.emplace_back(&v.certificate_, 4);
    for (const auto& v : a.governance_)
      objects.emplace_back(&v.certificate_, 4);
    std::map<ChunkKey, std::size_t> expected;
    std::map<Hash, ObjectRef> manifests;
    std::size_t remaining = 67108864;
    for (const auto& [v, kind] : objects) {
      take(validate_object(*v, kind));
      auto size = v->inline_.empty() ? v->reference_[0].byte_length_ : v->inline_.size();
      need(size <= remaining, "attachment-budget");
      remaining -= size;
      if (v->reference_.empty())
        continue;
      const auto& ref = v->reference_[0];
      auto [it, inserted] = manifests.emplace(ref.object_id_, ref);
      need(inserted || it->second == ref, "evidence-manifest-conflict");
      for (std::size_t i = 0; i < ref.chunk_hashes_.size(); ++i)
        expected.emplace(key(ref.object_id_, static_cast<std::uint8_t>(i)),
                         std::min(chunk_bytes, ref.byte_length_ - i * chunk_bytes));
    }
    // At most three objects with at most 64 chunks each. Iterate only as many
    // dictionary leaves as admitted by the typed manifests; never enumerate extras.
    vm::Dictionary supplied(present ? s.fetch_ref() : td::Ref<vm::Cell>{}, 264);
    vm::Dictionary canonical(264);
    std::map<ChunkKey, td::Ref<vm::Cell>> cells;
    need(supplied.check_for_each([&](td::Ref<vm::CellSlice> value, td::ConstBitPtr bits, int width) {
      need(width == 264 && value->size() == 0 && value->size_refs() == 1, "evidence-chunk-shape");
      ChunkKey k{};
      td::BitPtr(k.data()).copy_from(bits, 264);
      auto it = expected.find(k);
      need(it != expected.end(), "evidence-extra-chunk");
      auto cell = value->prefetch_ref();
      need(declared(cell, chunk_bytes) == it->second, "evidence-chunk-length");
      cells.emplace(k, cell);
      need(canonical.set_ref(td::ConstBitPtr(k.data()), 264, cell, vm::Dictionary::SetMode::Add),
           "evidence-dictionary");
      return true;
    }),
         "evidence-dictionary");
    need(cells.size() == expected.size(), "evidence-missing-chunk");
    auto supplied_root = supplied.get_root_cell(), canonical_root = canonical.get_root_cell();
    need(supplied_root.is_null() == canonical_root.is_null() &&
             (supplied_root.is_null() || supplied_root->get_hash() == canonical_root->get_hash()),
         "evidence-dictionary");
    need(take(charge(67108864 - remaining)), "evidence-charge");
    std::map<ChunkKey, Bytes> chunks;
    for (const auto& [k, cell] : cells)
      chunks.emplace(k, take(unpack_bytes(cell, expected.at(k))));
    NativeEvidence result(a, header, std::move(chunks));
    auto reader = result.reader();
    for (const auto& [v, kind] : objects)
      take(reader.resolve(*v, kind));
    return result;
  } catch (const Error& e) {
    return e;
  } catch (const vm::VmError&) {
    return Error{"evidence-cell"};
  } catch (const vm::VmVirtError&) {
    return Error{"evidence-pruned"};
  }
}
Result<Anchor> NativeEvidence::authenticate_owner(const NativeFinalizedHistory& history) const {
  if (authorizations_.owner_.size() != 1)
    return Error{"evidence-owner-absent"};
  const auto& expected = authorizations_.owner_[0].proof_.anchor_;
  auto actual = history.authenticate_header(expected.seqno_, header_);
  if (!actual.ok())
    return actual.error();
  if (actual.value() != expected)
    return Error{"evidence-owner-anchor"};
  return actual;
}
}  // namespace tos::auth
